File: music_ibook_flutter/lib/models/student_note_attempt.dart
import 'package:music_ibook_flutter/models/lesson_note.dart';

class StudentNoteAttempt {
  final int id;
  final int practiceSessionId;
  final int lessonNoteId;
  final String expectedNote;
  final String playedNote;
  final double expectedAtSecond;
  final double playedAtSecond;
  final double timingErrorMs;
  final bool isCorrectPitch;
  final bool isCorrectTiming;
  final bool isCorrect;
  final String createdAtUtc;
  final LessonNote? lessonNote;

  StudentNoteAttempt({
    required this.id,
    required this.practiceSessionId,
    required this.lessonNoteId,
    required this.expectedNote,
    required this.playedNote,
    required this.expectedAtSecond,
    required this.playedAtSecond,
    required this.timingErrorMs,
    required this.isCorrectPitch,
    required this.isCorrectTiming,
    required this.isCorrect,
    required this.createdAtUtc,
    this.lessonNote,
  });

  factory StudentNoteAttempt.fromJson(Map<String, dynamic> json) {
    return StudentNoteAttempt(
      id: json['id'] ?? 0,
      practiceSessionId: json['practiceSessionId'] ?? 0,
      lessonNoteId: json['lessonNoteId'] ?? 0,
      expectedNote: json['expectedNote'] ?? '',
      playedNote: json['playedNote'] ?? '',
      expectedAtSecond: (json['expectedAtSecond'] ?? 0).toDouble(),
      playedAtSecond: (json['playedAtSecond'] ?? 0).toDouble(),
      timingErrorMs: (json['timingErrorMs'] ?? 0).toDouble(),
      isCorrectPitch: json['isCorrectPitch'] ?? false,
      isCorrectTiming: json['isCorrectTiming'] ?? false,
      isCorrect: json['isCorrect'] ?? false,
      createdAtUtc: json['createdAtUtc'] ?? '',
      lessonNote: json['lessonNote'] != null
          ? LessonNote.fromJson(json['lessonNote'])
          : null,
    );
  }
}

File: Music-IBook-API/Music-IBook-API/Models/StudentNoteAttempt.cs
namespace Music_IBook_API.Models
{
    public class StudentNoteAttempt
    {
        public long Id { get; set; }

        public long PracticeSessionId { get; set; }

        public long LessonNoteId { get; set; }

        public string ExpectedNote { get; set; } = "";

        public string PlayedNote { get; set; } = "";

        public double ExpectedAtSecond { get; set; }

        public double PlayedAtSecond { get; set; }

        public double TimingErrorMs { get; set; }

        public bool IsCorrectPitch { get; set; }

        public bool IsCorrectTiming { get; set; }

        public bool IsCorrect { get; set; }

        public string JudgeResult { get; set; } = ""; // PERFECT, GOOD, LATE, WRONG, MISS

        public DateTime CreatedAtUtc { get; set; } = DateTime.UtcNow;

        public PracticeSession PracticeSession { get; set; } = null!;

        public LessonNote LessonNote { get; set; } = null!;
    }
}
